# Distorted HTML{Embed|Object}Element Blocked Properties (distorted-html-embed-object-element-blocked-properties)

For security the following `HTML{Embed|Object}Element` properties are prohibited
in Lightning Locker:
-   getSVGDocument

## Rule Details

Example of **incorrect** code:

```js
embed.getSVGDocument();
```
