// @locker/eslint-plugin-locker/dist/rules/distorted-document-blocked-properties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const eslint_rule_maker_1 = require("@locker/eslint-rule-maker");
const util_1 = require("../util");
exports.default = eslint_rule_maker_1.createRule({
    meta: {
        docs: {
            description: `disallow blocked Document properties`,
            category: 'Lightning Locker',
            recommended: true,
            url: util_1.docURL('distorted-document-blocked-properties'),
        },
        type: 'problem',
    },
    rule: {
        message: ({ pattern }) => `Document#${pattern.slice(9)} is prohibited in Lightning Locker.`,
        search: [
            'document.exitFullscreen',
            'document.fullscreen',
            'document.fullscreenElement',
            'document.fullscreenEnabled',
            'document.mozCancelFullScreen',
            'document.mozFullScreen',
            'document.mozFullScreenElement',
            'document.mozFullScreenEnabled',
            'document.onfullscreenchange',
            'document.onfullscreenerror',
            'document.onmozfullscreenchange',
            'document.onmozfullscreenerror',
            'document.onrejectionhandled',
            'document.onsecuritypolicyviolation',
            'document.onunhandledrejection',
            'document.releaseCapture',
            'document.releaseEvents',
            'document.webkitFullScreenKeyboardInputAllowed',
            'document.write',
            'document.writeln',
        ],
    },
});
