// @locker/eslint-plugin-locker/dist/rules/no-export-platform-resource-loader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const eslint_rule_maker_1 = require("@locker/eslint-rule-maker");
const util_1 = require("../util");
exports.default = eslint_rule_maker_1.createRule({
    meta: {
        docs: {
            description: `disallow export from 'lightning/platformResourceLoader'`,
            category: 'Lightning Locker',
            recommended: true,
            url: util_1.docURL('no-export-platform-resource-loader'),
        },
    },
    create(context) {
        function detect(node) {
            const { source } = node;
            if (source && source.value === 'lightning/platformResourceLoader') {
                context.report({
                    node,
                    message: `Exporting directly from 'lightning/platformResourceLoader' is not supported in Lightning Locker.`,
                });
            }
        }
        return {
            ExportAllDeclaration: detect,
            ExportNamedDeclaration: detect,
        };
    },
});
