// @locker/eslint-plugin-locker/dist/rules/uncompiled-empty-window-location.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const eslint_rule_maker_1 = require("@locker/eslint-rule-maker");
const util_1 = require("../util");
exports.default = eslint_rule_maker_1.createRule({
    meta: {
        docs: {
            description: 'window.location is empty',
            category: 'Lightning Locker',
            url: util_1.docURL('uncompiled-empty-window-location'),
        },
        type: 'problem',
    },
    rule: {
        message: 'window.location is empty in uncompiled Lightning Locker.',
        search: ['window.location'],
        onMatch: (data) => {
            const parent = eslint_rule_maker_1.astLib.getParent(data.node);
            return (
            // Match property access on `window.location`, e.g. `window.location.href`.
            eslint_rule_maker_1.astLib.getType(parent) === 'MemberExpression' ||
                // Match assignment on `window.location`, e.g. `window.location = value`.
                eslint_rule_maker_1.matchers.matchAsNonWritable(data));
        },
    },
});
