// @locker/eslint-plugin-locker/dist/rules/uncompiled-no-dynamic-import.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const eslint_rule_maker_1 = require("@locker/eslint-rule-maker");
const util_1 = require("../util");
exports.default = eslint_rule_maker_1.createRule({
    meta: {
        docs: {
            description: `disallow dynamic 'import'`,
            category: 'Lightning Locker',
            url: util_1.docURL('uncompiled-no-dynamic-import'),
        },
        type: 'problem',
    },
    create(context) {
        return {
            // We use 'ESTree.ImportDeclaration' instead of 'ESTree.ImportExpression'
            // to avoid the following type error:
            //   Type 'ImportExpression' is not assignable to type 'Node'.
            //   Types of property 'source' are incompatible.
            ImportExpression(node) {
                context.report({
                    node,
                    message: `Dynamic 'import' is unsupported in uncompiled Lightning Locker.`,
                });
            },
        };
    },
});
